pkg/config/new.go

```diff
@@ -0,0 +1,238 @@
+	// behavior observable and testable in unit tests.
+	additionalConfigs []string
+}
+
+// New returns a Config as described in the containers.conf(5) man page.
+func New(options *Options) (*Config, error) {
+	if options == nil {
+		options = &Options{}
+	}
+	cachedConfigMutex.Lock()
+	defer cachedConfigMutex.Unlock()
+	return newLocked(options)
+}
+
+// Default returns the default container config.  If no default config has been
+// set yet, a new config will be loaded by New() and set as the default one.
+// All callers are expected to use the returned Config read only.  Changing
+// data may impact other call sites.
+func Default() (*Config, error) {
+	cachedConfigMutex.Lock()
+	defer cachedConfigMutex.Unlock()
+	if cachedConfig != nil || cachedConfigError != nil {
+		return cachedConfig, cachedConfigError
+	}
+	cachedConfig, cachedConfigError = newLocked(&Options{SetDefault: true})
+	return cachedConfig, cachedConfigError
+}
+
+// A helper function for New() expecting the caller to hold the
+// cachedConfigMutex.
+func newLocked(options *Options) (*Config, error) {
+	// Start with the built-in defaults
+	config, err := defaultConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Now, gather the system configs and merge them as needed.
+	configs, err := systemConfigs()
+	if err != nil {
+		return nil, fmt.Errorf("finding config on system: %w", err)
+	}
+	for _, path := range configs {
+		// Merge changes in later configs with the previous configs.
+		// Each config file that specified fields, will override the
+		// previous fields.
+		if err = readConfigFromFile(path, config); err != nil {
+			return nil, fmt.Errorf("reading system config %q: %w", path, err)
+		}
+		logrus.Debugf("Merged system config %q", path)
+		logrus.Tracef("%+v", config)
+	}
+
+	modules, err := options.modules()
+	if err != nil {
+		return nil, err
+	}
+
+	options.additionalConfigs = append(options.additionalConfigs, modules...)
+
+	// The _OVERRIDE variable _must_ always win.  That's a contract we need
+	// to honor (for the Podman CI).
+	if path := os.Getenv(containersConfOverrideEnv); path != "" {
+		if _, err := os.Stat(path); err != nil {
+			return nil, fmt.Errorf("%s file: %w", containersConfOverrideEnv, err)
+		}
+		options.additionalConfigs = append(options.additionalConfigs, path)
+	}
+
+	// If the caller specified a config path to use, then we read it to
+	// override the system defaults.
+	for _, add := range options.additionalConfigs {
+		if add == "" {
+			continue
+		}
+		// readConfigFromFile reads in container config in the specified
+		// file and then merge changes with the current default.
+		if err := readConfigFromFile(add, config); err != nil {
+			return nil, fmt.Errorf("reading additional config %q: %w", add, err)
+		}
+		logrus.Debugf("Merged additional config %q", add)
+		logrus.Tracef("%+v", config)
+	}
+	config.addCAPPrefix()
+
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	if err := config.setupEnv(); err != nil {
```

Comment:
Doing this for modules sounds wrong to me, we should not alter the envs so late in the code path when we read modules.

Reply:
Where else should we?  The logic predates this PR but I think the place is correct.  At this place all configs are loaded and only then we should set the "final" engine.env.

Reply:
yes but not for modules???

Reply:
sorry I got a wrong understanding of how this should work, this seems fine to now

Reply:
No worries, I appreciate the feedback :+1: